### PR TITLE
Fortinet FortiGate major bug fixes and improvements

### PR DESCRIPTION
**Pull request by mhofert-r7**
### Description

Describe the proposed changes:

  - Fix issue where get_address_objects was returning a JSON blob. It now returns a custom object
  - Add new search params to get_address_objects. These will make the block host workflows much less error prone.
  - Add helper methods to make action easier to read and future action easier to write.
  - Update some example inputs that were misleading.
  - Modify output from get_address_objects to show subnet as a CIDR as the output from this API call is the only time a subnet is used all other API calls output CIDR's and expect CIDR's as input.
  - Normalize input for all actions that take an ipmask as input. All such actions will now accept IP's CIDR's and subnetmasks as input

### Testing

Any testing information goes here.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://komand.github.io/python/style.html)

- [ ] For dependencies, pin [OS package](https://komand.github.io/python/style.html#dockerfile) and [Python package](https://komand.github.io/python/style.html#requirements-txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://komand.github.io/python/sdk.html#sdk-versions) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://komand.github.io/python/error_handling.html#plugin-exceptions) and [ConnectionTestException](https://komand.github.io/python/error_handling.html#connection-exceptions)
- [ ] For logging, use [self.logger](https://komand.github.io/python/sdk.html#logging)
- [ ] For docs, use [changelog style](https://komand.github.io/python/style.html#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://komand.github.io/python/style.html#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

Test run with `50.50.50.72` as input

```
{
  "body": {
    "log": "rapid7/Fortinet FortiGate:5.0.0. Step name: create_address_object\n",
    "meta": {},
    "output": {
      "response_object": {
        "build": 1066,
        "http_method": "POST",
        "http_status": 200,
        "mkey": "50.50.50.72/32",
        "name": "address",
        "old_revision": "80931130e17d20386df1b024ab403932",
        "path": "firewall",
        "revision": "7d75fc2350eeadc1c7930d820ced5e30",
        "revision_changed": true,
        "serial": "FGVM02TM20001791",
        "status": "success",
        "vdom": "root",
        "version": "v6.2.3"
      },
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/fortinet_fortigate:5.0.0 run < tests/create_address_object.json
</summary>
</details>

### Run

<details>

Test run with `10.10.10.10` as input

```
{
  "body": {
    "log": "rapid7/Fortinet FortiGate:5.0.0. Step name: get_address_objects\n",
    "meta": {},
    "output": {
      "address_objects": [
        {
          "allow-routing": "disable",
          "cache-ttl": 0,
          "clearpass-spt": "unknown",
          "color": 0,
          "end-mac": "00:00:00:00:00:00",
          "fsso-group": [],
          "list": [],
          "name": "Example Subnet",
          "q_origin_key": "Example Subnet",
          "sdn-addr-type": "private",
          "start-mac": "00:00:00:00:00:00",
          "sub-type": "sdn",
          "subnet": "10.10.10.10/32",
          "tagging": [],
          "type": "ipmask",
          "uuid": "a6517c96-a1c6-51ea-8cac-0c31b932649b",
          "visibility": "enable"
        }
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/fortinet_fortigate:5.0.0 run < tests/get_address_objects.json
</summary>
</details>

### CODE REVIEW

File: fortinet_fortigate/plugin.spec.yaml [hunk 1]
```diff
@@ -689,12 +689,24 @@ actions:
...
+        description: Optional subnet to filter on
+        type: string
+        required: false
+        example: 198.51.100.100/32
+      fqdn_filter:
+        title: FQDN Filter
+        description: Optional FQDN to filter on
+        type: string
+        required: false
+        example: rapid7.com
```
Comment (jschipp-r7) [inline fortinet_fortigate/plugin.spec.yaml:704]:
```suggestion
        example: example.com
```

File: fortinet_fortigate/help.md [hunk 2]
```diff
@@ -642,11 +646,13 @@ _This plugin does not contain any triggers._
...
 
-In order for this plugin to communicate with the Fortigate firewall, the orchestrator's IP address must be added to the *trusted hosts* list for the API admin account.
-To accomplish this, log into the FortiGate firewall. Go to the System tab -> Administrator subtab and then select and edit the API admin.
+In order for this plugin to function the orchestrator's IP address must be added to the trusted hosts list for the API admin
+To do this log into the FortiGate firewall. Go to the System tab, Administrator subtab. Then select and edit the API admin.
 Add the orchestrator's IP address to the trusted hosts in CIDR form e.g. `198.51.100.100/32`
+
 # Version History
 
+* 5.0.0 - Improve input handling for all actions to allow IPs, CIDRs, and subnetmasks to all be handled correctly | fixed output of get_address_objects to return searchable data | Update get_address_objects to allow for additional search parameters
```
Comment (jschipp-r7) [inline fortinet_fortigate/help.md:655]:
```suggestion
* 5.0.0 - Improve input handling to allow IPs, CIDRs, and subnet masks in actions | Fix output of get_address_objects to return searchable data | Update get_address_objects to allow for additional search parameters
```

File: fortinet_fortigate/help.md [hunk 3]
```diff
@@ -642,11 +646,13 @@ _This plugin does not contain any triggers._
 
 ## Troubleshooting
 
-In order for this plugin to communicate with the Fortigate firewall, the orchestrator's IP address must be added to the *trusted hosts* list for the API admin account.
-To accomplish this, log into the FortiGate firewall. Go to the System tab -> Administrator subtab and then select and edit the API admin.
+In order for this plugin to function the orchestrator's IP address must be added to the trusted hosts list for the API admin
+To do this log into the FortiGate firewall. Go to the System tab, Administrator subtab. Then select and edit the API admin.
```
Comment (jschipp-r7) [inline fortinet_fortigate/help.md:650]:
```suggestion
To accomplish this, log into the FortiGate firewall. Go to the System tab -> Administrator subtab and then select and edit the API admin account.
```

File: fortinet_fortigate/help.md [hunk 4]
```diff
@@ -642,11 +646,13 @@ _This plugin does not contain any triggers._
 
 ## Troubleshooting
 
-In order for this plugin to communicate with the Fortigate firewall, the orchestrator's IP address must be added to the *trusted hosts* list for the API admin account.
-To accomplish this, log into the FortiGate firewall. Go to the System tab -> Administrator subtab and then select and edit the API admin.
+In order for this plugin to function the orchestrator's IP address must be added to the trusted hosts list for the API admin
```
Comment (jschipp-r7) [inline fortinet_fortigate/help.md:649]: The previous wording was better IMO

```suggestion
To accomplish this, log into the FortiGate firewall. Go to the System tab -> Administrator subtab and then select and edit the API admin.
```

File: fortinet_fortigate/help.md [hunk 5]
```diff
@@ -458,21 +458,25 @@ This action is used to get address objects.
...
+|fqdn_filter|string|None|False|Optional FQDN to filter on|None|rapid7.com|
+|name_filter|string|None|False|Optional name to filter on|None|MaliciousHost|
+|subnet_filter|string|None|False|Optional subnet to filter on|None|198.51.100.100/32|
 
 Example input:
 
 ```
 {
-  "name_filter": "198.51.100.100"
+  "fqdn_filter": "rapid7.com",
```
Comment (jschipp-r7) [inline fortinet_fortigate/help.md:469]:
```suggestion
  "fqdn_filter": "example.com",
```

File: fortinet_fortigate/help.md [hunk 6]
```diff
@@ -642,11 +646,13 @@ _This plugin does not contain any triggers._
 
 ## Troubleshooting
 
-In order for this plugin to communicate with the Fortigate firewall, the orchestrator's IP address must be added to the *trusted hosts* list for the API admin account.
 To accomplish this, log into the FortiGate firewall. Go to the System tab -> Administrator subtab and then select and edit the API admin.
+To accomplish this, log into the FortiGate firewall. Go to the System tab -> Administrator subtab and then select and edit the API admin account.
```
Comment (mrinehart-r7) [inline fortinet_fortigate/help.md:650]:
```suggestion
```
Duplicate

File: fortinet_fortigate/help.md [hunk 7]
```diff
@@ -642,11 +646,13 @@ _This plugin does not contain any triggers._
...
 Add the orchestrator's IP address to the trusted hosts in CIDR form e.g. `198.51.100.100/32`
+
 # Version History
 
+* 5.0.0 - Improve input handling to allow IPs, CIDRs, and subnet masks in actions | Fix output of get_address_objects to return searchable data | Update get_address_objects to allow for additional search parameters
```
Comment (mrinehart-r7) [inline fortinet_fortigate/help.md:655]:
```suggestion
* 5.0.0 - Improve input handling to allow IPs, CIDRs, and subnet masks in actions | Fix output of Get Address Objects action to return usable data | Update Get Address Objects action to allow for additional search parameters
```

File: fortinet_fortigate/icon_fortinet_fortigate/actions/get_address_objects/action.py [hunk 8]
```diff
@@ -36,4 +37,9 @@ def run(self, params={}):
         helper.http_errors(json_response, response.status_code)
 
         results = response.json().get("results")
+        for i in range(len(results)):
+            if results[i].get("subnet"):
+                subnet = helper.ipmaskConverter(results[i].get("subnet"))
+                results[i]["subnet"] = subnet
```
Comment (mrinehart-r7) [inline fortinet_fortigate/icon_fortinet_fortigate/actions/get_address_objects/action.py:43]: Why not just `for result in results` and then `if result.get("subnet")`?

File: fortinet_fortigate/icon_fortinet_fortigate/util/util.py [hunk 9]
```diff
@@ -97,3 +97,35 @@ def match_whitelist(self, address: str, whitelist: list) -> bool:
...
+                return type_
+            elif host[-1].isdigit() or host[-2].isdigit():
+                raise PluginException(cause="The host input appears to be an invalid IP or domain name.",
+                                      assistance="Ensure that the host input is a valid IP or domain.",
+                                      data=host)
+            type_ = "fqdn"
+            return type_
+        return type_
+
+    def ipmaskConverter(self, host: str) -> str:
```
Comment (mrinehart-r7) [inline fortinet_fortigate/icon_fortinet_fortigate/util/util.py:117]:
```suggestion
    def ipmask_converter(self, host: str) -> str:
```

File: fortinet_fortigate/icon_fortinet_fortigate/util/util.py [hunk 10]
```diff
@@ -97,3 +97,35 @@ def match_whitelist(self, address: str, whitelist: list) -> bool:
...
+        """Converts a IP or netmask into a CIDR"""
+        try:
+            host = ip_network(host)
+        except ValueError:
+            host = host.replace(" ", "/")
+            host = ip_network(host)
+        return str(host)
+
+    def netmaskConverter(self, host: str) -> str:
```
Comment (mrinehart-r7) [inline fortinet_fortigate/icon_fortinet_fortigate/util/util.py:126]:
```suggestion
    def netmask_converter(self, host: str) -> str:
```

File: fortinet_fortigate/plugin.spec.yaml [hunk 11]
```diff
@@ -689,12 +689,24 @@ actions:
...
+        description: Optional FQDN to filter on
+        type: string
+        required: false
+        example: example.com
     output:
       address_objects:
         title: Address Objects
         description: A list of address objects
-        type: '[]object'
+        type: '[]address_objects'
```
Comment (mrinehart-r7) [inline fortinet_fortigate/plugin.spec.yaml:709]:
```suggestion
        type: '[]address_object'
```
The type should be singular (update the type definition)

File: fortinet_fortigate/icon_fortinet_fortigate/util/util.py [hunk 12]
```diff
@@ -98,7 +98,7 @@ def match_whitelist(self, address: str, whitelist: list) -> bool:
 
         return False
 
-    def TypeFinder(self, host: str) -> str:
+    def Type_finder(self, host: str) -> str:
```
Comment (mrinehart-r7) [inline fortinet_fortigate/icon_fortinet_fortigate/util/util.py:101]:
```suggestion
    def type_finder(self, host: str) -> str:
```

**Pull request merged**